modules/homa/src/lib.rs

```diff
@@ -16,118 +16,1065 @@
+	#[pallet::storage]
+	#[pallet::getter(fn redeem_requests)]
+	pub type RedeemRequests<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, (Balance, bool), OptionQuery>;
+
+	/// The records of unbonding by AccountId.
+	///
+	/// Unbondings: double_map AccountId, ExpireEraIndex => UnbondingStakingCurrencyAmount
+	#[pallet::storage]
+	#[pallet::getter(fn unbondings)]
+	pub type Unbondings<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, T::AccountId, Twox64Concat, EraIndex, Balance, ValueQuery>;
+
+	/// The estimated staking reward rate per era on relaychain.
+	///
+	/// EstimatedRewardRatePerEra: value: Rate
+	#[pallet::storage]
+	#[pallet::getter(fn estimated_reward_rate_per_era)]
+	pub type EstimatedRewardRatePerEra<T: Config> = StorageValue<_, Rate, ValueQuery>;
+
+	/// Th maximum amount of bonded staking currency for a single sub on relaychain to obtain the
+	/// best staking rewards.
+	///
+	/// SoftBondedCapPerSubAccount: value: Balance
+	#[pallet::storage]
+	#[pallet::getter(fn soft_bonded_cap_per_sub_account)]
+	pub type SoftBondedCapPerSubAccount<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// Th staking amount of threshold to mint.
+	///
+	/// MintThreshold: value: Balance
+	#[pallet::storage]
+	#[pallet::getter(fn mint_threshold)]
+	pub type MintThreshold<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// Th liquid amount of threshold to redeem.
+	///
+	/// RedeemThreshold: value: Balance
+	#[pallet::storage]
+	#[pallet::getter(fn redeem_threshold)]
+	pub type RedeemThreshold<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// The rate of Homa drawn from the staking reward as commission.
+	/// The draw will be transfer to TreasuryAccount of Homa in liquid currency.
+	///
+	/// CommissionRate: value: Rate
+	#[pallet::storage]
+	#[pallet::getter(fn commission_rate)]
+	pub type CommissionRate<T: Config> = StorageValue<_, Rate, ValueQuery>;
+
+	/// The fixed fee rate for redeem request is fast matched.
+	///
+	/// FastMatchFeeRate: value: Rate
+	#[pallet::storage]
+	#[pallet::getter(fn fast_match_fee_rate)]
+	pub type FastMatchFeeRate<T: Config> = StorageValue<_, Rate, ValueQuery>;
+
+	/// The prefix of block numbers to bump local current era.
+	///
+	/// BumpEraPrefix: value: BlockNumber
+	#[pallet::storage]
+	#[pallet::getter(fn bump_era_prefix)]
+	pub type BumpEraPrefix<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	/// The internal of blocks numbers to bump local current era.
+	#[pallet::storage]
+	#[pallet::getter(fn bump_era_frequency)]
+	pub type BumpEraFrequency<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
 	#[pallet::pallet]
 	pub struct Pallet<T>(_);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if Self::should_bump_local_current_era(n) {
```

Comment:
I think we should use relay chain block number for better accuracy.
The relay chain number in `on_initialize` is the one from previous block, but should be fine.